pkg/traceql/ast_execute.go

```diff
@@ -0,0 +1,117 @@
+package traceql
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/tempo/pkg/util/log"
+)
+
+func (f SpansetFilter) matches(span Span) (bool, error) {
+	static, err := f.Expression.execute(span)
+	if err != nil {
+		level.Debug(log.Logger).Log("msg", "SpanSetFilter.matches failed", "err", err)
+		return false, err
+	}
+	if static.Type != TypeBoolean {
+		level.Debug(log.Logger).Log("msg", "SpanSetFilter.matches did not return a boolean", "err", err)
+		return false, fmt.Errorf("result of SpanSetFilter (%v) is %v", f, static.Type)
+	}
+	return static.B, nil
+}
+
+func (o BinaryOperation) execute(span Span) (Static, error) {
+	lhs, err := o.LHS.execute(span)
+	if err != nil {
+		return NewStaticNil(), err
+	}
+
+	rhs, err := o.RHS.execute(span)
+	if err != nil {
+		return NewStaticNil(), err
+	}
+
+	// TODO validate types are valid similar to validate()
+
+	switch o.Op {
+	// TODO implement arithmetics
+	case OpAdd:
+	case OpSub:
+	case OpDiv:
+	case OpMod:
+	case OpMult:
+	case OpGreater:
+		return NewStaticBool(lhs.asFloat() > rhs.asFloat()), nil
+	case OpGreaterEqual:
+		return NewStaticBool(lhs.asFloat() >= rhs.asFloat()), nil
+	case OpLess:
+		return NewStaticBool(lhs.asFloat() < rhs.asFloat()), nil
+	case OpLessEqual:
+		return NewStaticBool(lhs.asFloat() <= rhs.asFloat()), nil
+	case OpPower:
+	case OpEqual:
+		return NewStaticBool(lhs.Equals(rhs)), nil
+	case OpNotEqual:
+		return NewStaticBool(!lhs.Equals(rhs)), nil
+	case OpRegex:
+		matched, err := regexp.MatchString(rhs.S, lhs.S)
```

Comment:
Should this return false when both arguments aren't strings? There is a failing test that covers this, `3 =~ 5` which returns true because both `.S` are empty.  

Same applies to `OpNotRegex`.

Reply:
Yeah, it should. I removed some code right before pushing that validates `lhs` and `rhs` have the same type. It's the same code as `validate()` but filling in the the types we got from the storage layer. Once I get those back in the test should be okay again.

Reply:
Latest commit has extra logic to validate the types after they have been resolved with span data. This is the same logic as `validate()`, just using data from the storage layer.